src/Simulation/Simulators/SparseSimulator/SparseSimulatorCS/SparseSimulator.cs

```diff
@@ -0,0 +1,388 @@
+            {
+                throw new Exception(msg);
+            }
+        }
+
+        [DllImport(simulator_dll)] 
+        private static extern double JointEnsembleProbability_cpp(uint sim, int length, int[] basis, int[] qubits);
+        public override void AssertProb(IQArray<Pauli> bases, IQArray<Qubit> qubits, double probabilityOfZero, string msg, double tol)
+        {
+            double result = JointEnsembleProbability_cpp(Id, bases.Count(), bases.Select(x => (int)x).ToArray(), qubits.Select(x => x.Id).ToArray());
+            if (System.Math.Abs(1.0 - result - probabilityOfZero) > tol)
+            {
+                var extendedMsg = $"{msg}\n\tExpected:\t{probabilityOfZero}\n\tActual:\t{result}";
+                IgnorableAssert.Assert(false, extendedMsg);
+                throw new ExecutionFailException(extendedMsg);
+            }
+        }
+
+
+        [DllImport(simulator_dll)]
+        private static extern void allocateQubit_cpp(uint sim, int qubit_id);
+        public override void OnAllocateQubits(IQArray<Qubit> qubits)
+        {
+            try
+            {
+                foreach (Qubit qubit in qubits)
+                {
+                    allocateQubit_cpp(Id, qubit.Id);
+                }
+            } catch (ExternalException ex)
+            {
+                // Possibly C++ errors: qubit already occupied, or not enough qubits
+                throw ex;
+            }
+        }
+
+        [DllImport(simulator_dll)]
+        private static extern void releaseQubit_cpp(uint sim, int qubit_id);
+        public override void OnReleaseQubits(IQArray<Qubit> qubits)
+        {
+            try
+            {
+                foreach (Qubit qubit in qubits)
+                {
+                    releaseQubit_cpp(Id, qubit.Id);
+                }
+            } catch (ExternalException)
+            {
+                // Possible C++ errors: Qubit not in zero state
+                throw new Microsoft.Quantum.Simulation.Simulators.Exceptions.ReleasedQubitsAreNotInZeroState();
+            }
+        }
+
+        // Tells C++ to delete any objects, to manage memory
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        [DllImport(simulator_dll)]
+        private static extern void destroy_cpp(uint sim);
+        public void Dispose(bool Disposing)
+        {
+            destroy_cpp(Id);
+        }
+
+        public override void OnOperationStart(ICallable operation, IApplyData arguments)
+        {
+            // Writes the operation to console if it is logging
+            if (isLogging)
+            {
+                Console.WriteLine(stackDepth + operation.FullName + "{");
+                stackDepth += " ";
+            }
+        }
+
+        public override void OnOperationEnd(ICallable operation, IApplyData arguments)
+        {
+            
+            if (isLogging)
+            {
+                stackDepth = stackDepth[0..(stackDepth.Length - 1)];
```

Comment:
I don't know C# but I have an impression that this line does not change anything.